tsl/src/chunk.c

```diff
@@ -259,3 +262,60 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type)
 
 	return num_results;
 }
+
+static bool
+chunk_is_distributed(const Chunk *chunk)
+{
+	return chunk->relkind == RELKIND_FOREIGN_TABLE;
+}
+
+Datum
+chunk_create_replica_table(PG_FUNCTION_ARGS)
+{
+	Oid chunk_relid;
+	const char *data_node_name;
+	const Chunk *chunk;
+	const Hypertable *ht;
+	const ChunkDataNode *cdn;
+	Cache *hcache = ts_hypertable_cache_pin();
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	GETARG_NOTNULL_OID(chunk_relid, 0, "chunk");
+	GETARG_NOTNULL_NULLABLE(data_node_name, 1, "data node name", CSTRING);
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, false);
+	if (chunk == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("relation \"%s\" is not a chunk", get_rel_name(chunk_relid))));
+	if (!chunk_is_distributed(chunk))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("chunk \"%s\" doesn't belong to a distributed hypertable",
+						get_rel_name(chunk_relid))));
+
+	ht = ts_hypertable_cache_get_entry(hcache, chunk->hypertable_relid, CACHE_FLAG_NONE);
+	ts_hypertable_permissions_check(ht->main_table_relid, GetUserId());
+
+	/* Check the given data node exists */
+	data_node_get_foreign_server(data_node_name, ACL_USAGE, true, false);
+	/* Find if hypertable is attached to the data node and return an error otherwise */
+	data_node_hypertable_get(ht, data_node_name, true);
+
+	cdn = ts_chunk_data_node_scan_by_chunk_id_and_node_name(chunk->fd.id,
```

Comment:
The `chunk` object has a list of data nodes `chunk->data_nodes` already. I suggest checking against that list of nodes since that would avoid the extra scan.

Reply:
Do you know if there is already a helper function for this check?

Reply:
I created one